src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+            for (uint8_t i = 0; i < d->NumberOfModes(); i++)
+            {
+                ModeOptionStructType mode;
+                ChipError err1 = CHIP_NO_ERROR;
+
+                // Get the mode label
+                char buffer[64];
+                MutableCharSpan label(buffer);
+                err1 = d->GetModeLabelByIndex(i, label);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+                mode.label = label;
+
+                // Get the mode value
+                err1 = d->GetModeValueByIndex(i, mode.mode);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+
+                // Get the mode tags
+                ModeTagStructType tagsBuffer[8];
+                List<ModeTagStructType> tags(tagsBuffer);
+                err1 = d->GetModeTagsByIndex(i, tags);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+                mode.modeTags = tags;
+
+                ReturnErrorOnFailure(encoder.Encode(mode));
+            }
+            return CHIP_NO_ERROR;
+        });
+        ReturnErrorOnFailure(err);
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+// Implements checking before attribute writes.
+CHIP_ERROR Instance::Write(const ConcreteDataAttributePath & attributePath, AttributeValueDecoder & aDecoder)
+{
+    DataModel::Nullable<uint8_t> newMode;
+    ReturnErrorOnFailure(aDecoder.Decode(newMode));
+
+    if (newMode.IsNull()) // This indicates that the new mode is null.
+    {
+        switch (attributePath.mAttributeId)
+        {
+        case ModeBase::Attributes::StartUpMode::Id:
+            UpdateStartUpMode(newMode);
+            return CHIP_NO_ERROR;
+        case ModeBase::Attributes::OnMode::Id:
+            UpdateOnMode(newMode);
+            return CHIP_NO_ERROR;
+        }
+    }
+    else
+    {
+        if (!IsSupportedMode(newMode.Value()))
+        {
+            return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+        }
+
+        switch (attributePath.mAttributeId)
+        {
+        case ModeBase::Attributes::StartUpMode::Id:
+            UpdateStartUpMode(newMode);
+            return CHIP_NO_ERROR;
+        case ModeBase::Attributes::OnMode::Id:
+            UpdateOnMode(newMode);
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+}
+
+void Instance::UpdateCurrentMode(uint8_t aNewMode)
+{
+    uint8_t oldMode = mCurrentMode;
+    mCurrentMode    = aNewMode;
+    if (mCurrentMode != oldMode)
+    {
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::CurrentMode::Id);
```

Comment:
What about having the implementation given do something with the mode when it has changed? Is it up to the overridding method to do a duplicate check of mode check, and THEN call the base class method? Seems necessary otherwise the edge is missed. Would have been great to have a delegate method called on these edges.

Reply:
When a request to change the current mode attribute via the command is made, the delegate's (will be a delegate again soon) handle function is called and if successfully handled, the server will write the new value. I have raised an [discussion](https://csamembers.slack.com/archives/CUXQ5QRMK/p1688046344943149) about changing the spec so that we implement the same patter for attribute reads.  So at the moment, according to the spec, when an attribute write is requested, the implementation does not have a way to reject it.

I don't think that an implementation would need to do anything else when the OnMode or StartUpMode is changed, other than possibly rejecting the write.

Reply:
> when an attribute write is requested, the implementation does not have a way to reject it

Actually, a write can be rejected:

> 8.7.3.3. Write Path Data Process
> If the path indicates a fabric-scoped list or list entry, it SHALL be processed as a fabric-filtered list of fabric-scoped structs.
> 
> Each data field indicated by the path, SHALL be processed in the order conveyed as follows:
> 
> If a data field is not within the constraints defined by the cluster specification, an AttributeStatusIB SHALL be generated with the CONSTRAINT_ERROR Status Code, with the Path field duplicating the path.
> 
> Otherwise perform the following:
> 
> The data field SHALL be changed to the data indicated with the path.
> 
> An AttributeStatusIB SHALL be generated with the SUCCESS Status Code, with the Path field duplicating the path.

Reply:
Correct. According to this, the only way a write can be rejected is if the value is invalid, in this case that means that it is not a valid mode. This is already dealt with in the write. It has now been imported in these update functions.

However, apart from that, there is noting else for the delegate to do here, except for the CurrentMode which it would have already handled to get here.